File: depertin_cliente/test/constants/tipos_entrega_test.dart
// Arquivo: test/constants/tipos_entrega_test.dart
//
// Testes unitários puros da classe `TiposEntrega` (sem Firebase). Cobre:
//  - normalização de lista vinda do Firestore
//  - cálculo do maior tipo (base de frete)
//  - compatibilidade entrega × loja (usada no filtro de despacho)
//  - default conservador para lojas legado
//  - cadeia de fallback de tabela de frete
//  - leitura direta de mapa (lerDeDoc)
//
// Estes testes protegem o contrato compartilhado entre client, painel web e
// Cloud Functions: os 3 builds implementam a mesma lógica em linguagens
// diferentes (Dart × Dart × Node). Qualquer mudança aqui precisa ser
// replicada em `depertin_web/lib/constants/tipos_entrega.dart` e
// `depertin_cliente/functions/tipos_entrega.js`.

import 'package:flutter_test/flutter_test.dart';
import 'package:depertin_cliente/constants/tipos_entrega.dart';

void main() {
  group('TiposEntrega.normalizarLista', () {
    test('descarta valores fora do set canônico', () {
      final r = TiposEntrega.normalizarLista(
        <dynamic>['moto', 'drone', 'carro', null, 123, 'BICICLETA'],
      );
      expect(r, <String>['bicicleta', 'moto', 'carro']);
    });

    test('remove duplicatas e aplica lower-case', () {
      final r = TiposEntrega.normalizarLista(
        <dynamic>['MOTO', 'moto', 'Moto', 'carro', 'carro'],
      );
      expect(r, <String>['moto', 'carro']);
    });

    test('ordena por hierarquia ascendente (bike→moto→carro→frete)', () {
      final r = TiposEntrega.normalizarLista(
        <dynamic>['carro_frete', 'bicicleta', 'carro', 'moto'],
      );
      expect(r, <String>['bicicleta', 'moto', 'carro', 'carro_frete']);
    });

    test('retorna lista vazia para null ou não-iterável', () {
      expect(TiposEntrega.normalizarLista(null), const <String>[]);
      expect(TiposEntrega.normalizarLista(42), const <String>[]);
      expect(TiposEntrega.normalizarLista('moto'), const <String>[]);
    });
  });

  group('TiposEntrega.maiorTipoDaLista', () {
    test('retorna null para lista vazia', () {
      expect(TiposEntrega.maiorTipoDaLista(const <String>[]), isNull);
    });

    test('moto < carro < carro_frete', () {
      expect(
        TiposEntrega.maiorTipoDaLista(const ['bicicleta', 'moto']),
        'moto',
      );
      expect(
        TiposEntrega.maiorTipoDaLista(const ['moto', 'carro']),
        'carro',
      );
      expect(
        TiposEntrega.maiorTipoDaLista(
          const ['bicicleta', 'moto', 'carro', 'carro_frete'],
        ),
        'carro_frete',
      );
    });

    test('retorna o único item quando a lista tem tamanho 1', () {
      expect(
        TiposEntrega.maiorTipoDaLista(const ['bicicleta']),
        'bicicleta',
      );
    });
  });

  group('TiposEntrega.compativel (filtro de despacho)', () {
    test('loja sem config (legado) aceita qualquer entregador', () {
      expect(TiposEntrega.compativel('moto', const <String>[]), isTrue);
      expect(TiposEntrega.compativel('bicicleta', const <String>[]), isTrue);
      expect(TiposEntrega.compativel('', const <String>[]), isTrue);
    });

    test('loja que só aceita carro_frete filtra motoboy', () {
      expect(
        TiposEntrega.compativel('moto', const ['carro_frete']),
        isFalse,
      );
      expect(
        TiposEntrega.compativel('carro_frete', const ['carro_frete']),
        isTrue,
      );
    });

    test('loja com múltiplos tipos aceita qualquer um da lista', () {
      const aceitos = ['bicicleta', 'moto', 'carro'];
      expect(TiposEntrega.compativel('bicicleta', aceitos), isTrue);
      expect(TiposEntrega.compativel('moto', aceitos), isTrue);
      expect(TiposEntrega.compativel('carro', aceitos), isTrue);
      expect(TiposEntrega.compativel('carro_frete', aceitos), isFalse);
    });
  });

  group('TiposEntrega.defaultLegado (Opção B — conservador)', () {
    test('sem produto volumoso → apenas [moto] (preserva pré-migração)', () {
      expect(
        TiposEntrega.defaultLegado(temProdutoRequerVeiculoGrande: false),
        const <String>['moto'],
      );
    });

    test('com produto volumoso → [carro, carro_frete]', () {
      expect(
        TiposEntrega.defaultLegado(temProdutoRequerVeiculoGrande: true),
        const <String>['carro', 'carro_frete'],
      );
    });

    test('bicicleta NÃO entra em default (opt-in explícito do lojista)', () {
      final sem = TiposEntrega.defaultLegado(
        temProdutoRequerVeiculoGrande: false,
      );
      final com = TiposEntrega.defaultLegado(
        temProdutoRequerVeiculoGrande: true,
      );
      expect(sem, isNot(contains('bicicleta')));
      expect(com, isNot(contains('bicicleta')));
    });
  });

  group('TiposEntrega.cadeiaFallbackTabela', () {
    test('carro_frete → carro_frete → carro → padrao', () {
      expect(
        TiposEntrega.cadeiaFallbackTabela[TiposEntrega.codCarroFrete],
        const <String>['carro_frete', 'carro', 'padrao'],
      );
    });

    test('carro → carro → padrao', () {
      expect(
        TiposEntrega.cadeiaFallbackTabela[TiposEntrega.codCarro],
        const <String>['carro', 'padrao'],
      );
    });

    test('moto e bicicleta caem em padrao diretamente', () {
      expect(
        TiposEntrega.cadeiaFallbackTabela[TiposEntrega.codMoto],
        const <String>['padrao'],
      );
      expect(
        TiposEntrega.cadeiaFallbackTabela[TiposEntrega.codBicicleta],
        const <String>['padrao'],
      );
    });
  });

  group('TiposEntrega.lerDeDoc', () {
    test('retorna vazio para null', () {
      expect(TiposEntrega.lerDeDoc(null), const <String>[]);
    });

    test('retorna vazio se o campo não existir', () {
      expect(
        TiposEntrega.lerDeDoc(<String, dynamic>{'outra_coisa': true}),
        const <String>[],
      );
    });

    test('lê e normaliza quando campo existe', () {
      final r = TiposEntrega.lerDeDoc(<String, dynamic>{
        'tipos_entrega_permitidos': <dynamic>['CARRO', 'moto', 'bicicleta'],
      });
      expect(r, const <String>['bicicleta', 'moto', 'carro']);
    });

    test('respeita o parâmetro `campo` customizado', () {
      final r = TiposEntrega.lerDeDoc(
        <String, dynamic>{'meu_campo': <dynamic>['moto']},
        campo: 'meu_campo',
      );
      expect(r, const <String>['moto']);
    });
  });

  group('TiposEntrega.paraFirestore', () {
    test('normaliza antes de persistir (evita lixo no Firestore)', () {
      final r = TiposEntrega.paraFirestore(
        const <String>['CARRO', 'moto', 'moto', 'drone'],
      );
      expect(r, const <String>['moto', 'carro']);
    });
  });

  group('TiposEntrega.normalizarTipoVeiculo (paridade com backend)', () {
    test('retorna vazio para null, string vazia e lixo', () {
      expect(TiposEntrega.normalizarTipoVeiculo(null), '');
      expect(TiposEntrega.normalizarTipoVeiculo(''), '');
      expect(TiposEntrega.normalizarTipoVeiculo('   '), '');
      expect(TiposEntrega.normalizarTipoVeiculo('drone'), '');
      expect(TiposEntrega.normalizarTipoVeiculo(42), '');
    });

    test('reconhece códigos canônicos exatos', () {
      expect(TiposEntrega.normalizarTipoVeiculo('bicicleta'), 'bicicleta');
      expect(TiposEntrega.normalizarTipoVeiculo('moto'), 'moto');
      expect(TiposEntrega.normalizarTipoVeiculo('carro'), 'carro');
      expect(
        TiposEntrega.normalizarTipoVeiculo('carro_frete'),
        'carro_frete',
      );
    });

    test('normaliza variações com maiúsculas e espaços', () {
      expect(TiposEntrega.normalizarTipoVeiculo('MOTO'), 'moto');
      expect(TiposEntrega.normalizarTipoVeiculo(' Carro '), 'carro');
      expect(TiposEntrega.normalizarTipoVeiculo('Bicicleta'), 'bicicleta');
    });

    test('frete tem prioridade sobre carro (Fiorino, Kombi, pick-up, van)', () {
      expect(
        TiposEntrega.normalizarTipoVeiculo('Fiorino'),
        'carro_frete',
      );
      expect(TiposEntrega.normalizarTipoVeiculo('Kombi'), 'carro_frete');
      expect(
        TiposEntrega.normalizarTipoVeiculo('Pick-up'),
        'carro_frete',
      );
      expect(
        TiposEntrega.normalizarTipoVeiculo('Van de carga'),
        'carro_frete',
      );
      expect(
        TiposEntrega.normalizarTipoVeiculo('carro de frete'),
        'carro_frete',
      );
      expect(
        TiposEntrega.normalizarTipoVeiculo('Utilitário'),
        'carro_frete',
      );
      expect(
        TiposEntrega.normalizarTipoVeiculo('utilitario'),
        'carro_frete',
      );
    });

    test('bike / bicy / bicicleta viram bicicleta', () {
      expect(TiposEntrega.normalizarTipoVeiculo('Bike'), 'bicicleta');
      expect(TiposEntrega.normalizarTipoVeiculo('bicy'), 'bicicleta');
      expect(
        TiposEntrega.normalizarTipoVeiculo('Bicicleta elétrica'),
        'bicicleta',
      );
    });

    test('moto / scooter / motocicleta viram moto', () {
      expect(TiposEntrega.normalizarTipoVeiculo('Scooter'), 'moto');
      expect(
        TiposEntrega.normalizarTipoVeiculo('Motocicleta'),
        'moto',
      );
      expect(TiposEntrega.normalizarTipoVeiculo('Motoboy'), 'moto');
    });
  });

  group('Hierarquia canônica (regra de negócio)', () {
    test('bicicleta(1) < moto(2) < carro(3) < carro_frete(4)', () {
      expect(TiposEntrega.hierarquia['bicicleta'], 1);
      expect(TiposEntrega.hierarquia['moto'], 2);
      expect(TiposEntrega.hierarquia['carro'], 3);
      expect(TiposEntrega.hierarquia['carro_frete'], 4);
    });

    test('tabela de frete mapeia corretamente', () {
      expect(TiposEntrega.tabelaFretePorTipo['bicicleta'], 'padrao');
      expect(TiposEntrega.tabelaFretePorTipo['moto'], 'padrao');
      expect(TiposEntrega.tabelaFretePorTipo['carro'], 'carro');
      expect(TiposEntrega.tabelaFretePorTipo['carro_frete'], 'carro_frete');
    });

    test('raio recomendado bicicleta ≤ 2km (limite operacional)', () {
      final raioBike = TiposEntrega.raioKmRecomendado['bicicleta'];
      expect(raioBike, isNotNull);
      expect(raioBike, lessThanOrEqualTo(2.0));
    });
  });
}

File: depertin_cliente/functions/test/tipos_entrega.test.js
// Arquivo: functions/test/tipos_entrega.test.js
//
// Smoke tests do helper Node `functions/tipos_entrega.js`. Roda com o
// test runner nativo do Node 18+ (zero dependências adicionais):
//
//   npm test   (após adicionar `"test": "node --test test/"` ao package.json)
//   ou
//   node --test test/tipos_entrega.test.js
//
// Este arquivo é o espelho JS dos testes Dart em
// `depertin_cliente/test/constants/tipos_entrega_test.dart` — garantem que
// os 3 builds (mobile, web, backend) concordem sobre a regra de negócio.

"use strict";

const test = require("node:test");
const assert = require("node:assert/strict");
const T = require("../tipos_entrega");

test("normalizarLista descarta valores fora do canônico", () => {
    assert.deepEqual(
        T.normalizarLista(["moto", "drone", "carro", null, 123, "BICICLETA"]),
        ["bicicleta", "moto", "carro"],
    );
});

test("normalizarLista remove duplicatas e lowercase", () => {
    assert.deepEqual(
        T.normalizarLista(["MOTO", "moto", "Moto", "carro", "carro"]),
        ["moto", "carro"],
    );
});

test("normalizarLista ordena por hierarquia ascendente", () => {
    assert.deepEqual(
        T.normalizarLista(["carro_frete", "bicicleta", "carro", "moto"]),
        ["bicicleta", "moto", "carro", "carro_frete"],
    );
});

test("normalizarLista devolve [] para não-array", () => {
    assert.deepEqual(T.normalizarLista(null), []);
    assert.deepEqual(T.normalizarLista(42), []);
    assert.deepEqual(T.normalizarLista("moto"), []);
});

test("maiorTipo: null para vazio", () => {
    assert.equal(T.maiorTipo([]), null);
    assert.equal(T.maiorTipo(null), null);
});

test("maiorTipo: escolhe o topo da hierarquia", () => {
    assert.equal(T.maiorTipo(["moto", "carro"]), "carro");
    assert.equal(
        T.maiorTipo(["bicicleta", "moto", "carro", "carro_frete"]),
        "carro_frete",
    );
    assert.equal(T.maiorTipo(["bicicleta"]), "bicicleta");
});

test("compativel: loja sem config aceita qualquer entregador (legado)", () => {
    assert.equal(T.compativel("moto", []), true);
    assert.equal(T.compativel("", []), true);
    assert.equal(T.compativel("bicicleta", null), true);
});

test("compativel: loja só carro_frete filtra motoboy", () => {
    assert.equal(T.compativel("moto", ["carro_frete"]), false);
    assert.equal(T.compativel("carro_frete", ["carro_frete"]), true);
});

test("compativel: loja com múltiplos tipos aceita todos da lista", () => {
    const aceitos = ["bicicleta", "moto", "carro"];
    assert.equal(T.compativel("bicicleta", aceitos), true);
    assert.equal(T.compativel("moto", aceitos), true);
    assert.equal(T.compativel("carro", aceitos), true);
    assert.equal(T.compativel("carro_frete", aceitos), false);
});

test("defaultLegado (Opção B): sem volumoso → [moto]", () => {
    assert.deepEqual(
        T.defaultLegado({ temProdutoRequerVeiculoGrande: false }),
        ["moto"],
    );
});

test("defaultLegado: com volumoso → [carro, carro_frete]", () => {
    assert.deepEqual(
        T.defaultLegado({ temProdutoRequerVeiculoGrande: true }),
        ["carro", "carro_frete"],
    );
});

test("defaultLegado: bicicleta nunca entra em default", () => {
    const sem = T.defaultLegado({ temProdutoRequerVeiculoGrande: false });
    const com = T.defaultLegado({ temProdutoRequerVeiculoGrande: true });
    assert.ok(!sem.includes("bicicleta"));
    assert.ok(!com.includes("bicicleta"));
});

test("CADEIA_FALLBACK_TABELA.carro_frete cobre 3 níveis", () => {
    assert.deepEqual(
        T.CADEIA_FALLBACK_TABELA[T.COD.CARRO_FRETE],
        ["carro_frete", "carro", "padrao"],
    );
});

test("CADEIA_FALLBACK_TABELA.carro cai em padrao", () => {
    assert.deepEqual(
        T.CADEIA_FALLBACK_TABELA[T.COD.CARRO],
        ["carro", "padrao"],
    );
});

test("CADEIA_FALLBACK_TABELA moto e bike usam padrao direto", () => {
    assert.deepEqual(T.CADEIA_FALLBACK_TABELA[T.COD.MOTO], ["padrao"]);
    assert.deepEqual(T.CADEIA_FALLBACK_TABELA[T.COD.BICICLETA], ["padrao"]);
});

test("HIERARQUIA bike(1)<moto(2)<carro(3)<frete(4)", () => {
    assert.equal(T.HIERARQUIA.bicicleta, 1);
    assert.equal(T.HIERARQUIA.moto, 2);
    assert.equal(T.HIERARQUIA.carro, 3);
    assert.equal(T.HIERARQUIA.carro_frete, 4);
});

test("lerDeDoc: null → []", () => {
    assert.deepEqual(T.lerDeDoc(null), []);
    assert.deepEqual(T.lerDeDoc(undefined), []);
});

test("lerDeDoc lê tipos_entrega_permitidos e normaliza", () => {
    assert.deepEqual(
        T.lerDeDoc({ tipos_entrega_permitidos: ["CARRO", "moto", "bicicleta"] }),
        ["bicicleta", "moto", "carro"],
    );
});

test("lerDeDoc: doc sem campo → []", () => {
    assert.deepEqual(T.lerDeDoc({ outra_coisa: true }), []);
});

test("normalizarTipoVeiculo: variações comuns", () => {
    assert.equal(T.normalizarTipoVeiculo("Moto"), "moto");
    assert.equal(T.normalizarTipoVeiculo("motocicleta"), "moto");
    assert.equal(T.normalizarTipoVeiculo("scooter"), "moto");
    assert.equal(T.normalizarTipoVeiculo("Carro popular"), "carro");
    assert.equal(T.normalizarTipoVeiculo("carro"), "carro");
    assert.equal(T.normalizarTipoVeiculo("carro_frete"), "carro_frete");
    assert.equal(T.normalizarTipoVeiculo("Fiorino"), "carro_frete");
    assert.equal(T.normalizarTipoVeiculo("pick-up"), "carro_frete");
    assert.equal(T.normalizarTipoVeiculo("Kombi"), "carro_frete");
    assert.equal(T.normalizarTipoVeiculo("utilitário"), "carro_frete");
    assert.equal(T.normalizarTipoVeiculo("Bike"), "bicicleta");
    assert.equal(T.normalizarTipoVeiculo("bicicleta"), "bicicleta");
    assert.equal(T.normalizarTipoVeiculo(""), "");
    assert.equal(T.normalizarTipoVeiculo(null), "");
    assert.equal(T.normalizarTipoVeiculo("avião"), "");
});

test("normalizarTipoSolicitado: aceita código canônico direto", () => {
    assert.equal(T.normalizarTipoSolicitado("moto"), "moto");
    assert.equal(T.normalizarTipoSolicitado("carro"), "carro");
    assert.equal(T.normalizarTipoSolicitado("carro_frete"), "carro_frete");
    assert.equal(T.normalizarTipoSolicitado("bicicleta"), "bicicleta");
});

test("normalizarTipoSolicitado: aceita variações livres (ex: Carro Popular)", () => {
    assert.equal(T.normalizarTipoSolicitado("Carro Popular"), "carro");
    assert.equal(T.normalizarTipoSolicitado("motocicleta"), "moto");
    assert.equal(T.normalizarTipoSolicitado("Pick-up"), "carro_frete");
});

test("normalizarTipoSolicitado: retorno vazio em lixo/null", () => {
    assert.equal(T.normalizarTipoSolicitado(null), "");
    assert.equal(T.normalizarTipoSolicitado(""), "");
    assert.equal(T.normalizarTipoSolicitado("drone"), "");
    assert.equal(T.normalizarTipoSolicitado(42), "");
});

test("categoriaEfetivaPedido: tipo_entrega_solicitado explícito sobrescreve lista aceita", () => {
    const r = T.categoriaEfetivaPedido(
        { tipo_entrega_solicitado: "moto" },
        ["moto", "carro"],
    );
    assert.equal(r, "moto");
});

test("categoriaEfetivaPedido: loja com 1 tipo aceito → tipo implícito", () => {
    const r = T.categoriaEfetivaPedido({}, ["moto"]);
    assert.equal(r, "moto");
});

test("categoriaEfetivaPedido: loja multi-tipo sem escolha → null", () => {
    const r = T.categoriaEfetivaPedido({}, ["moto", "carro"]);
    assert.equal(r, null);
});

test("categoriaEfetivaPedido: loja sem config e pedido sem escolha → null", () => {
    assert.equal(T.categoriaEfetivaPedido({}, []), null);
    assert.equal(T.categoriaEfetivaPedido({}, null), null);
});

test("categoriaEfetivaPedido: variação 'Carro Popular' no solicitado é canonizada", () => {
    const r = T.categoriaEfetivaPedido(
        { tipo_entrega_solicitado: "Carro Popular" },
        ["moto", "carro"],
    );
    assert.equal(r, "carro");
});

test("compativel: escolha explícita de moto não admite entregador de carro", () => {
    assert.equal(T.compativel("moto", ["moto"]), true);
    assert.equal(T.compativel("carro", ["moto"]), false);
    assert.equal(T.compativel("moto", ["carro"]), false);
});
